processor/processor.go

```diff
@@ -0,0 +1,37 @@
+// Copyright 2019, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package processor
+
+import (
+	"context"
+
+	"github.com/census-instrumentation/opencensus-service/data"
+)
+
+// MetricsDataProcessor is an interface that receives data.MetricsData, process it as needed, and
+// sends it to the next processing node if any or to the destination.
+//
+// ProcessMetricsData receives data.MetricsData for processing by the MetricsDataProcessor.
+type MetricsDataProcessor interface {
+	ProcessMetricsData(ctx context.Context, md data.MetricsData) error
+}
+
+// TraceDataProcessor is an interface that receives data.TraceData, process it as needed, and
+// sends it to the next processing node if any or to the destination.
```

Comment:
If it sends it to the next destination, we probably need a `RegisterNext(...)` or something

Reply:
Not really, I was thinking that we will add that to the factory method that creates the Processor. Reason being that you cannot update the next stages at runtime.

Reply:
I was also thinking we could potentially have this interface that packages need to implement, and then a struct that wraps this interface that deals with the common "core" functionality, such as reconfiguring (reloading) the TraceDataProcessor from its factory, and storing the `name` of the instance.

Reply:
Technically it doesn't need to send to anything after it. It makes sense to have this interface as it is to be a consumer in the pipeline and a separate interface to indicate that it sends its output (if any) to others (it could be part of creation but perhaps a separate interface). The example being the final blocks in the pipeline (exporters): they should have these Process(Metrics|Trace)Data but they don't have a next one.